File: apps/mybookkeeper/backend/app/schemas/properties/utility_plan_summary.py
"""Summary row for the utility-plan list view.

``property_name``, ``renewal_status`` and ``days_until_term_end`` are joined /
derived by the service — they are not columns. See
``utility_plan_service.renewal_status``.
"""
from __future__ import annotations

import datetime as _dt
import uuid
from decimal import Decimal

from pydantic import BaseModel, ConfigDict


class UtilityPlanSummary(BaseModel):
    id: uuid.UUID
    property_id: uuid.UUID
    property_name: str | None = None
    service_type: str
    provider_name: str
    plan_name: str | None = None
    rate_type: str
    avg_price_cents_per_kwh_at_1000: Decimal | None = None
    term_end_date: _dt.date | None = None
    # Negative once the term has lapsed, which is the case worth seeing.
    days_until_term_end: int | None = None
    renewal_status: str
    is_current: bool
    created_at: _dt.datetime
    updated_at: _dt.datetime

    model_config = ConfigDict(from_attributes=True)

File: apps/mybookkeeper/frontend/src/shared/types/utility/utility-plan-summary.ts
import type { UtilityPlanRateType } from "@/shared/types/utility/utility-plan-rate-type";
import type { UtilityPlanRenewalStatus } from "@/shared/types/utility/utility-plan-renewal-status";
import type { UtilityServiceType } from "@/shared/types/utility/utility-service-type";

/**
 * Summary row for the utility-plan list.
 *
 * Mirrors ``schemas/properties/utility_plan_summary.py``.
 *
 * Rate columns are ``Numeric`` on the backend and serialize as JSON *strings*
 * (``"13.9000"``), not numbers — parse before doing arithmetic.
 */
export interface UtilityPlanSummary {
  id: string;
  property_id: string;
  property_name: string | null;
  service_type: UtilityServiceType;
  provider_name: string;
  plan_name: string | null;
  rate_type: UtilityPlanRateType;
  avg_price_cents_per_kwh_at_1000: string | null;
  term_end_date: string | null;
  /** Negative once the term has lapsed, which is the case worth seeing. */
  days_until_term_end: number | null;
  renewal_status: UtilityPlanRenewalStatus;
  is_current: boolean;
  created_at: string;
  updated_at: string;
}
